### PR TITLE
Add Check Merchant Manager Role Middleware enhancement

### LINKED ISSUE #22: Add Check Merchant Manager Role Middleware
Create a middleware to check the merchant manager role.

### PR DESCRIPTION
**Pull request by ManasJadhav**
Creates a RequiredAuthorities annotation that takes in two paramerters.

1. requiredAuthorities array - roles required to access the endpoint.
2.source - source to get the merchantId (header, params, body)

### CODE REVIEW

File: .DS_Store
Comment (rohitvpatil0810) [inline .DS_Store:1]: ignore this files - add in .gitignore

File: src/.DS_Store
Comment (rohitvpatil0810) [inline src/.DS_Store:1]: ignore this file - add in .gitignore

File: src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java:1]: add here field to check option - like field String type - e.g. merchantId
replace array of MerchantRole with MerchantRole only

File: src/main/java/com/whytowait/core/annotations/UniqueValue.java
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/UniqueValue.java:1]: no change needed in file - revert

File: src/main/java/com/whytowait/core/annotations/ValidEnum.java
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/ValidEnum.java:1]: no change needed in file - revert

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        if (activeMerchantClaim == null || activeMerchantClaim.isEmpty()) {
+            throw new RuntimeException(new BadRequestException("MerchantId not found in request"));
+        }
+
+        if (!hasRequiredAuthorities(requiredAuthorities, activeMerchantClaim)) {
+            throw new RuntimeException(new UnauthorizedException("Required role not found"));
+        }
+    }
+
+    private String GetActiveMerchantClaim(RequestSource source) throws UnauthorizedException, BadRequestException {
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: Rename this function to something else - doesn't match with operation performed in function - rename to extract fieldValue something like this and use proper case for function name

File: src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.whytowait.core.annotations.enums.RequestSource;
+import com.whytowait.domain.models.enums.MerchantRole;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.METHOD, ElementType.TYPE})
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java]: Rename this interface to CheckMerchantManagerRole

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import java.util.Arrays;
+import java.util.List;
+
+@Aspect
+@Component
+public class RequiresAuthoritiesAspect {
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: Rename aspect class name to CheckMerchantManagerRoleAspect

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    @Before("@annotation(requiresAuthorities)")
+    public void handleRequiresAuthorities(RequiresAuthorities requiresAuthorities) throws UnauthorizedException, BadRequestException {
+
+        MerchantRole[] requiredAuthorities = requiresAuthorities.requiredAuthorities();
+        RequestSource source = requiresAuthorities.source();
+
+        String activeMerchantClaim = GetActiveMerchantClaim(source);
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: name of variable should be extractedFieldValue, something like this.

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    ;
+
+    public boolean hasRequiredAuthorities(MerchantRole[] requiredAuthorities, String merchantId) {
+
+        List<MerchantRole> requiredAuths = Arrays.stream(requiredAuthorities).toList();
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        List<String> authorities = authentication.getAuthorities().stream().map(GrantedAuthority::getAuthority).toList();
+
+        if (requiredAuths.contains(MerchantRole.MERCHANT_OWNER)) {
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: change this if-else logic to a number-based map, and check for role seniority based on number only. 
Not readable code.

File: src/main/java/com/whytowait/core/annotations/enums/RequestSource.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.whytowait.core.annotations.enums;
+
+import lombok.Getter;
+
+@Getter
+public enum RequestSource {
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/enums/RequestSource.java:6]: don't provide fieldName here, give the option to put fieldName to check in the annotation itself.

File: src/main/java/com/whytowait/core/annotations/enums/RequestSource.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    HEADER("merchant-id"),
+    PARAMS("merchantId"),
+    BODY("merchantId");
+
+    private final String source;
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/enums/RequestSource.java]: remove source, add source to annotation definition

File: src/main/java/com/whytowait/core/annotations/validators/EnumValidator.java
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/validators/EnumValidator.java:1]: no changes in the file - revert back.

File: src/main/java/com/whytowait/core/annotations/validators/UniqueValueValidator.java
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/validators/UniqueValueValidator.java:1]: no changes in the file - revert back.

**Review by rohitvpatil0810 — CHANGES_REQUESTED** [review]
Please make the necessary changes and resubmit.

File: src/main/java/com/whytowait/core/annotations/validators/UniqueValueValidator.java
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/validators/UniqueValueValidator.java:1]: Added that to directory, imports are changed

File: src/main/java/com/whytowait/core/annotations/validators/EnumValidator.java
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/validators/EnumValidator.java:1]: Added to directoy imports are changed

File: src/main/java/com/whytowait/core/annotations/enums/RequestSource.java (hunk 7 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/enums/RequestSource.java]: Source is added to annotation defination only, this is used for getting the field value.

File: src/main/java/com/whytowait/core/annotations/enums/RequestSource.java (hunk 6 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/enums/RequestSource.java:6]: What other fields do we need to check, we need only merchantId. i don't think we need user to tell us.

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java (hunk 1 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: done

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java (hunk 4 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: done

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java (hunk 3 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: done

File: src/main/java/com/whytowait/core/annotations/ValidEnum.java
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/ValidEnum.java:1]: Added to directory, imports are changed

File: src/main/java/com/whytowait/core/annotations/UniqueValue.java
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/UniqueValue.java:1]: Added to directoy, imports are changed

File: src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java (hunk 2 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java]: done

File: src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/RequiresAuthorities.java:1]: done

File: src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java (hunk 5 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/aspects/RequiresAuthoritiesAspect.java]: done

File: src/main/java/com/whytowait/core/annotations/CheckMerchantManagerRole.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import java.lang.annotation.Target;
+
+@Target({ElementType.METHOD, ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface CheckMerchantManagerRole {
+
+    MerchantRole requiredAuthority();
+
+    RequestSource source();
+
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/CheckMerchantManagerRole.java:20]: Here add the fieldName String type - field and you can also set default value of this field Name.
Change name of requiredAuthority to requiredRole

e.g. usage will be `@CheckMerchantManagerRole(source = RequestSource.PARAM, requiredRole = MerchantRole.MERCHANT_MANAGER, fieldName = "merchId")`

File: src/main/java/com/whytowait/core/annotations/enums/RequestSource.java (hunk 6 above)
Comment (rohitvpatil0810) [reply src/main/java/com/whytowait/core/annotations/enums/RequestSource.java:6]: The user will not tell us. We will use this in our code; we need to define in which field we expect the value to be checked for resource crud operations.

**Review by rohitvpatil0810 — CHANGES_REQUESTED** [review]
Please find comments and reiterate.

File: src/main/java/com/whytowait/core/annotations/aspects/CheckMerchantManagerRoleAspect.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        // Get the current user's authorities
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        List<String> authorities = authentication.getAuthorities()
+                .stream()
+                .map(GrantedAuthority::getAuthority)
+                .toList();
+
+        // Check roles priority-wise
+        String userRole = "MERCHANT_" + merchantId + "_" + requiredAuthority.name();
+        return authorities.contains(userRole);
```
Comment (rohitvpatil0810) [inline src/main/java/com/whytowait/core/annotations/aspects/CheckMerchantManagerRoleAspect.java]: earlier implementation was correct only, I was just asking to rewrite it in number based checking or role.

Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/aspects/CheckMerchantManagerRoleAspect.java]: done

File: src/main/java/com/whytowait/core/annotations/enums/RequestSource.java (hunk 6 above)
Comment (ManasJadhav) [reply src/main/java/com/whytowait/core/annotations/enums/RequestSource.java:6]: done added a fieldvalue which will tell us where to look for the property.

**Review by rohitvpatil0810 — APPROVED** [review]

**Pull request merged**